executor/slow_query.go

```diff
@@ -588,346 +586,143 @@ func (e *slowQueryRetriever) parseLog(ctx context.Context, sctx sessionctx.Conte
-		record = append(record, types.NewIntDatum(1))
-	} else {
-		record = append(record, types.NewIntDatum(0))
+func (e *slowQueryRetriever) setDefaultValue(row []types.Datum) {
+	for i := range row {
+		if !row[i].IsNull() {
+			continue
+		}
+		row[i] = table.GetZeroValue(e.outputCols[i])
 	}
-	if st.planFromBinding {
-		record = append(record, types.NewIntDatum(1))
-	} else {
-		record = append(record, types.NewIntDatum(0))
+}
+
+type slowQueryColumnValueFactory func(row []types.Datum, value string, tz *time.Location, checker *slowLogChecker) (valid bool, err error)
+
+func parseUserOrHostValue(value string) string {
+	// the new User&Host format: root[root] @ localhost [127.0.0.1]
+	tmp := strings.Split(value, "[")
+	return strings.TrimSpace(tmp[0])
+}
+
+func getColumnValueFactoryByName(sctx sessionctx.Context, colName string, columnIdx int) (slowQueryColumnValueFactory, error) {
+	switch colName {
+	case variable.SlowLogTimeStr:
+		return func(row []types.Datum, value string, tz *time.Location, checker *slowLogChecker) (bool, error) {
+			t, err := ParseTime(value)
+			if err != nil {
+				return false, err
+			}
+			timeValue := types.NewTime(types.FromGoTime(t), mysql.TypeTimestamp, types.MaxFsp)
+			if checker != nil {
+				valid := checker.isTimeValid(timeValue)
+				if !valid {
+					return valid, nil
+				}
+			}
+			if t.Location() != tz {
+				t = t.In(tz)
+				timeValue = types.NewTime(types.FromGoTime(t), mysql.TypeTimestamp, types.MaxFsp)
+			}
+			row[columnIdx] = types.NewTimeDatum(timeValue)
+			return true, nil
+		}, nil
+	case variable.SlowLogBackoffDetail:
+		return func(row []types.Datum, value string, tz *time.Location, checker *slowLogChecker) (bool, error) {
+			backoffDetail := row[columnIdx].GetString()
+			if len(backoffDetail) > 0 {
+				backoffDetail += " "
+			}
+			backoffDetail += value
+			row[columnIdx] = types.NewStringDatum(backoffDetail)
+			return true, nil
+		}, nil
+	case variable.SlowLogPlan:
+		return func(row []types.Datum, value string, tz *time.Location, checker *slowLogChecker) (bool, error) {
+			plan := parsePlan(value)
+			row[columnIdx] = types.NewStringDatum(plan)
+			return true, nil
+		}, nil
+	case variable.SlowLogConnIDStr, variable.SlowLogExecRetryCount, variable.SlowLogPreprocSubQueriesStr,
+		execdetails.WriteKeysStr, execdetails.WriteSizeStr, execdetails.PrewriteRegionStr, execdetails.TxnRetryStr,
+		execdetails.RequestCountStr, execdetails.TotalKeysStr, execdetails.ProcessKeysStr,
+		execdetails.RocksdbDeleteSkippedCountStr, execdetails.RocksdbKeySkippedCountStr,
+		execdetails.RocksdbBlockCacheHitCountStr, execdetails.RocksdbBlockReadCountStr,
+		variable.SlowLogTxnStartTSStr, execdetails.RocksdbBlockReadByteStr:
+		return func(row []types.Datum, value string, tz *time.Location, checker *slowLogChecker) (valid bool, err error) {
+			v, err := strconv.ParseUint(value, 10, 64)
+			if err != nil {
+				return false, err
+			}
+			row[columnIdx] = types.NewUintDatum(v)
+			return true, nil
```

Comment:
do we have some utility function that can append row according to field type?

Reply:
No. 
Actually, the other system table will append all column's values into row and do prune later.

Reply:
Then how about extracting this into a utility function? Seems that this can be useful for other system tables.

Reply:
No, this is not suitable for other system tables. Since only `SLOW_QUERY` should parse `string` from `slow log`, but the other system table no need to do parse.